### PR TITLE
[SYSTEMML-2057] Support bitwise operators and, or, xor, & LShift, Rshift

### PR DESCRIPTION
**Pull request by j143**
_What is here?_
Cleanly added 
1. `bitwAnd`
2. `bitwOr`
3. `bitwXor`
4. `bitwShiftL`
5. `bitwShiftR`
these operations in dml, can be supported over all three combinations(?), unlike `.R`.

Implemented as:
```
parser [DMLTranslator, BuiltinFunctionExpression] - > Hops[Hop, BinaryOp] - > Lops[Binary, BinaryScalar] -> @runtime ->Instructions -> [CP, MR, SP] -> fnobject ->execute()
```

### CODE REVIEW

**Comment by j143** [comment]
Hi, this patch is not complete (haven't rebased too!), After remaining tests & `not` implementation I will complete this.

Also, I am trying to figure out, how to run test suite with R enabled. But, will test manually.
Thanks.

**Comment by akchinSTC** [comment]
Build failed, see build log for details

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2228/<h2>Failed Tests: <span class='status-failure'>3</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2228/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>3</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2228/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2228/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor1/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2228/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.scalar/XorTest/testXor4/'><strong>org.apache.sysml.test.integration.functions.binary.scalar.XorTest.testXor4</strong></a></li></ul>

**Comment by j143** [comment]
Hi @mboehm7 , thanks a lot for consolidation to `Binary`. Here, I have added all, but `bitwNot` will follow tomorrow.

1. I could'nt run unit test, but in `scripts/test.dml` file I have written cleanly the results.
2. there is some problem with `matrix-scalar`.
Thanks.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/BinaryScalarScalarCPInstruction.java [hunk 1]
```diff
@@ -65,10 +65,14 @@ else if( so1 instanceof DoubleObject || so2 instanceof DoubleObject || output.ge
 				sores = new DoubleObject( dop.fn.execute(so1.getDoubleValue(), so2.getDoubleValue()) );
 			}
 			else if( so1 instanceof IntObject || so2 instanceof IntObject ) {
```
Comment (j143) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/BinaryScalarScalarCPInstruction.java]: if it is `bitwAnd(1, 2.0)`, then this is being evaluated here itself, but can it be on the next loop.

cc @mboehm7

**Comment by mboehm7** [comment]
yes that's a good idea. For all bitwise operations, we should explicitly cast all inputs to integer, which is in line with R semantics (from the R docs: "numeric vectors are coerced to integer vectors"). Please use {{UtilFunctions.toLong(x)}} for safe casts which avoids anomalies if an integer cannot be represented exactly in double.

**Comment by akchinSTC** [comment]
Build finished.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/<h2>Failed Tests: <span class='status-failure'>30</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>30</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2231/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndDenseCP</strong></a></li></ul>

**Comment by j143** [comment]
Hi @mboehm7 , 
1. anyway we could only support double & would it be fine (I believe), until we support integer computation to keep it as double only. or should we just give chance for the both?

2. BTW, how can I run tests with `R`, locally. How to download & integrate `R`, with **IntelliJ IDEA**

**Comment by mboehm7** [comment]
1) I would recommend to do the cast in the individual function objects' execute methods `public double execute(double in1, double in2)` - that way you can leave the scalar instruction as is, and it will work for both scalars and matrices with the exact semantics of R.

2) You can simply run your test as a junit test inside your IDE. If you get errors that Rscript cannot be found, you need to put the path of the bin directory of your R installation into the global PATH variable (OS env variable).

**Comment by mboehm7** [comment]
@j143 do you wanna make the change above or do you need help?

**Comment by j143** [comment]
I am trying to install R to test on my computer, & will be reaching as soon as possible after testing.

Thanks.

**Comment by j143** [comment]
Hi @mboehm7 , In the `scripts/test.dml`, file I have kept the results (I've problems with R setup!). However, `matrix-scalar` is not working.

Thanks.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/<h2>Failed Tests: <span class='status-failure'>30</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>30</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwOrDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwOrDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwAndSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwAndSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftRSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftRSparseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLDenseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLDenseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLDenseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLDenseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwShiftLDenseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwShiftLDenseSP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorSparseCP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorSparseCP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorSparseMR/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorSparseMR</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2237/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.binary.matrix/ElementwiseBitwLogicalTest/testBitwXorSparseSP/'><strong>org.apache.sysml.test.integration.functions.binary.matrix.ElementwiseBitwLogicalTest.testBitwXorSparseSP</strong></a></li></ul>

**Comment by mboehm7** [comment]
ok, then let me have a look what's going on here and merge this once the remaining issues are fixed. Thanks @j143.

**Comment by mboehm7** [comment]
ok thanks again @j143 - here is just a quick summary of what was missing
* Fix of R and dml scripts (e.g., license in one R script, conversion to vectors)
* Cleanup of function objects, with support for long-long, double-double inputs (now with proper int semantics to match R results)
* Add missing scalar-scalar and matrix-scalar compiler and runtime support (incl fixes for the existing AND, OR, XOR operators)

From a style perspective, in the future, please leave the individual imports - using the wildcard makes it harder to resolve if there are multiple wildcards.

**Comment by j143** [comment]
Hi Matthias,

Sorry about the wildcard import that seems to be automatic in my ide, I will make sure that I will follow your code style. I will go through the changeset once more, so that things will be clearer to me. 

Thanks.

**Pull request closed without merging**